### ISSUE TITLE
Mail templates

### ISSUE THREAD

**Issue by daviskregers** [rainlab/translate-plugin#279]
Hello,

Are there any plans on adding a mail template translation option?

Currently I tried extending the model:

```
                \System\Models\MailTemplate::extend(function($model){

                    $model->implement[] = 'RainLab.Translate.Behaviors.TranslatableModel';

                    $model->translatable[] = 'subject';

                    $model->translatable[] = 'description';

                    $model->translatable[] = 'content_html';

                });
```

But the method doesn't seem to be working, it throws an error:

`"Indirect modification of overloaded property System\Models\MailTemplate::$translatable has no effect"`

**Comment by LukeTowers** [comment]
Try adding it as a property first:
```php
\System\Models\MailTemplate::extend(function($model){
    $model->addDynamicProperty('translateable', ['subject', 'description', 'content_html']);
}
```
http://octobercms.com/docs/services/behaviors#constructor-extension

**Comment by pvullioud** [comment]
Any idea why twig translation is not working in email template. Other Twig function are working fine.
What could be the best way to fix it ?

**Comment by daviskregers** [comment]
Hello,

Thanks for the reply. It doesn't seem to be working though - the addDynamicProperty throws an exception:

Unexpected type of array, should attribute "translatable" be jsonable?
vendor/october/rain/src/Database/Model.php line 1127

The $model->jsonable seems to be returning null, if I change it to ['translatable], it doesn't seem to be taking any effect. If I use the same addDynamicProperty it throws the same error on the jsonable field.

**Comment by LukeTowers** [comment]
Any progress on this @daviskregers?

**Comment by rashArt** [comment]
hi @LukeTowers ,

I have the same case, only that it works in the fields description and plaintext, i changed from type: codeeditor to textarea in the field content_html does not take the translation, is there anything that prevents the other elements from taking the translation?

This is my call to rainlab.translate:
`MailTemplateModel::extend(function ($model){`
`$model->implement[] = 'RainLab.Translate.Behaviors.TranslatableModel';`
`$model->addDynamicProperty('translatable', ['subject', 'description', 'content_html', 'content_text']);`
`});`

**Comment by munxar** [comment]
Have the exact same problem. If i define the translatable property in the MailTemplate model itself, it works:
```php
class MailTemplate extends Model
{
    use \October\Rain\Database\Traits\Validation;
    public $translatable = [];
    ...
```

if I use the addDynamicProperty approach, it displays the translatable fields, but on save it gives me the error:
`Unexpected type of array, should attribute "translatable" be jsonable?" on line 1127 of vendor/october/rain/src/Database/Model.php`

sidenote: the subject is defined as 'subject@create' and 'subject@update' to have different displays on create and update. I had to add exactly this notation to the $translate array in order to make it work for 'subject'

**Comment by LukeTowers** [comment]
Not a clue @munxar, you're going to have to do some more digging on this one. Let me know how it goes!

**Comment by munxar** [comment]
@LukeTowers right now it seams that the only difference is that if declared directly in the class, it isn't serialized, if added dynamic it is.
but why ...

**Comment by LukeTowers** [comment]
Take a look at the extension base classes in the library, see if you can figure it out

**Comment by munxar** [comment]
Finally I solved it:

```php
MailTemplate::extend(function($model) {
            $model->extendClassWith(TranslatableModel::class);
            $model->addDynamicProperty('translatable', ['subject@create', 'subject@update', 'content_html', 'content_text']);
});
```

Using **extendClassWith()** instead of setting the $implements property did the trick.

**Comment by LukeTowers** [comment]
@munxar how well does the mail template translation work? Does it know when to send out the correct templates?

**Comment by munxar** [comment]
@LukeTowers Not, yet. I'm on it. The Idea is that Backend\User and User have a communication language (or not, fallback default locale) associated.
Now the hard part is to intercept the Mail facade, but due to the existing events, it should work.
Any hints or ideas?

**Comment by LukeTowers** [comment]
I know there are some mailer events that you could think about hooking into: https://docs.google.com/spreadsheets/d/17vv4zn_ezp9iFBgYhYuYsmUG4MXW9uuIw7aOIrVcHiM/edit#gid=0. I'm interested to see what you come up with though :)

**Comment by Samuell1** [comment]
Hey @munxar , did you got it working? Trying your extend function after sending email i get error `ErrorException: Invalid argument supplied for foreach() in ...\plugins\rainlab\translate\classes\TranslatableBehavior.php:314`

**Comment by munxar** [comment]
no, stopped investigating. sorry.

**Comment by mjauvin** [comment]
Closing as it has been over a month since any activity on this occurred and we are trying to figure out what issues are still relevant. If this is still something that you would like to see through to fruition please respond and we can get the ball rolling.

**Comment by adrian-marinescu-ch** [comment]
@mjauvin I encounter same issue by trying to translate the `MailTemplate` model. Has anyone found a solution for this?

**Comment by mjauvin** [comment]
@adrian-marinescu-ch If I add the following, it seems to work just fine:
```
        \System\Models\MailTemplate::extend(function ($model) {
            $model->addDynamicProperty('translatable', []);
            $model->translatable = array_merge($model->translatable, ['subject@create', 'subject@update', 'description', 'content_html', 'content_text']);
            $model->extendClassWith('RainLab\Translate\Behaviors\TranslatableModel');
        }); 
```

**Comment by adrian-marinescu-ch** [comment]
@mjauvin I somehow figured out, first **you need to have e-mail templates imported in the database**: `\System\Models\MailTemplate::syncAll();` and only after that you can make the model translatable. If not, you'll get `$jsonable` error. I synced the mail templates on seed and now everything works just fine. This is my extend code:
```
        MailTemplate::extend(function ($model) {
            if (Settings::get('is_seeded', false)) {
                $model->implement[] = 'RainLab.Translate.Behaviors.TranslatableModel';
                $model->addDynamicProperty(
                    'translatable',
                    [
                      'subject@create',
                      'subject@update',
                      'description',
                      'content_html',
                      'content_text'
                    ]
                );
            }
        });
```

**Comment by mjauvin** [comment]
@adrian-marinescu-ch do you get the same behavior if you use the exact following code?
```
     MailTemplate::extend(function ($model) {
            $model->addDynamicProperty('translatable', []);
            $model->translatable = array_merge($model->translatable, ['subject@create', 'subject@update', 'description', 'content_html', 'content_text']);
            $model->extendClassWith('October\Rain\Database\Behaviors\Purgeable');
            $model->extendClassWith('RainLab\Translate\Behaviors\TranslatableModel');
});
```

**Comment by AIC-BV** [comment]
> ```
>         \System\Models\MailTemplate::extend(function ($model) {
>             $model->addDynamicProperty('translatable', []);
>             $model->translatable = array_merge($model->translatable, ['subject@create', 'subject@update', 'description', 'content_html', 'content_text']);
>             $model->extendClassWith('RainLab\Translate\Behaviors\TranslatableModel');
>         }); 
> ```

Thank you, I  managed to translated my mail templates using the code above in the boot() function!
Just one small issue, even though the mail is translatable, the mail always sends in the default language and doesn't listen to the activeLocale..

Any ideas how to solve this problem so that the mail sends in the correct language?

**Comment by daftspunk** [comment]
@AIC-BV, what code are you using to send the mail?

**Comment by AIC-BV** [comment]
`// set the mail template
        $mailTemplate = 'aic.contact::mail.contact';

        // remove empty fields
        $data = array_filter($data);

        // replace '-' with '_'
        // because '-' doesn't work in twig langauge
        $data = $this->replaceArrayKeys($data);
        
        // send mail with a queue
        Mail::queue($mailTemplate, $data, function($message) use (&$data) {

            // get the theme
            $theme = Theme::getEditTheme();

            // get the theme variables
            $sender = $theme->mail_email;
            $senderName = $theme->mail_name;
            $bccEnabled = $theme->bcc_enabled;

            // receiver
            $email = $data['email'];
            $name = $data['fname'] . ' ' . $data['lname'];

            // set properties
            $message->from($sender, $senderName);
            $message->to($email, $name);

            // add bcc
            if ($bccEnabled) {

                // get theme variables
                $bccEmail = $theme->bcc_email;
                $bccName = $theme->bcc_name;

                // set bcc email
                if ($bccEmail && $bccEmail != '') {
                    $message->bcc($bccEmail, $bccName);
                }

            }

        });`

**Comment by AIC-BV** [comment]
> @AIC-BV, what code are you using to send the mail?

`// set the mail template
        $mailTemplate = 'aic.contact::mail.contact';

        // remove empty fields
        $data = array_filter($data);

        // replace '-' with '_'
        // because '-' doesn't work in twig langauge
        $data = $this->replaceArrayKeys($data);
        
        // send mail with a queue
        Mail::queue($mailTemplate, $data, function($message) use (&$data) {

            // get the theme
            $theme = Theme::getEditTheme();

            // get the theme variables
            $sender = $theme->mail_email;
            $senderName = $theme->mail_name;
            $bccEnabled = $theme->bcc_enabled;

            // receiver
            $email = $data['email'];
            $name = $data['fname'] . ' ' . $data['lname'];

            // set properties
            $message->from($sender, $senderName);
            $message->to($email, $name);

            // add bcc
            if ($bccEnabled) {

                // get theme variables
                $bccEmail = $theme->bcc_email;
                $bccName = $theme->bcc_name;

                // set bcc email
                if ($bccEmail && $bccEmail != '') {
                    $message->bcc($bccEmail, $bccName);
                }

            }

        });`

**Comment by daftspunk** [comment]
Thanks @AIC-BV 

This commit should fix it in the next release (v2.1.12)

https://github.com/octobercms/library/commit/3ad1eb378e66fffc24b5ef0c8f1836f4d28ab222

**Comment by AIC-BV** [comment]
Hey I'm sorry!
This is all very confusing to me, but I'm glad that you found a solution for the original OctoberCMS!
